0.39->0.40 CLI `migrate` command

amaurymartiny opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #6751 

What I did: checked out tag `v0.39.0`, looked into all `x/**/types/genesis.go` structs , and compared them to latest master's genesis structs (defined in proto files).

Also:
- The module-specific `Migrate` functions now take a `client.Context`. It's only use in 0.40, to decide which JSON (amino or proto) to output.

Update needed for:
- auth: BaseAccount has an Any pubkey field
- evidence: remove a Params field
- supply: merged into bank
- slashing: SigningInfos & ValidatorMissedBlocks move from map to array

If you can think of other genesis state's that I might have forgot, please shout.

depends on:
- [x] #7268
- [x] #6881 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

amaurymartiny commented on proto/cosmos/distribution/genesis.proto:
```diff
@@ -116,10 +116,10 @@ message ValidatorSlashEventRecord {
   uint64 height = 2;
   uint64 period = 3;
-  ValidatorSlashEvent event = 4 [
+  ValidatorSlashEvent validator_slash_event = 4 [
```
reason: the json field was `validator_slash_event` in v0.39

https://github.com/cosmos/cosmos-sdk/blob/89ceb62a463e783950d559fbe7164613565edae5/x/distribution/types/genesis.go#L51

fedekunze commented on x/auth/legacy/v0_38/types.go:
```diff
@@ -223,6 +223,10 @@ func NewBaseVestingAccount(
 }
 
 func (bva BaseVestingAccount) Validate() error {
+	fmt.Printf("%+v\n", bva)
+	fmt.Println("bva.Coins.IsZero()", bva.Coins.IsZero())
+	fmt.Println("bva.OriginalVesting.IsZero()", bva.OriginalVesting.IsZero())
+	fmt.Println("OriginalVesting=", bva.OriginalVesting, "Coins=", bva.Coins)
```
```suggestion
```

alexanderbez commented on x/bank/legacy/v0_40/migrate.go:
```diff
@@ -2,21 +2,33 @@ package v040
+	balances := make([]types.Balance, len(authGenState.Accounts))
 	for i, acc := range authGenState.Accounts {
-		balances[i] = Balance{
+		balances[i] = types.Balance{
```
migration logic cannot rely on anything in modules. You'll have to copy `Balance` to `types.go` in this pkg.

alexanderbez commented on x/bank/legacy/v0_40/migrate.go:
```diff
@@ -2,21 +2,33 @@ package v040
+		types.NewParams(bankGenState.SendEnabled, types.SendEnabledParams{}),
+		balances,
+		supplyGenState.Supply,
+		[]types.Metadata{},
```
ditto

alexanderbez commented on x/bank/legacy/v0_40/types.go:
```diff
@@ -3,41 +3,6 @@ package v040
 	ModuleName = "bank"
 )
-
-var _ GenesisBalance = (*Balance)(nil)
```
undo. The reason we copy the types is that they need to be pinned to their respective version, whereas if we simply reference the `types` package, they can change.

amaurymartiny commented on x/evidence/legacy/v0_40/migrate_test.go:
```diff
@@ -0,0 +1,46 @@
+  ]
+}`
+
+	bz, err := v040Codec.MarshalJSONIndent(migrated, "", "  ")
```
I have an error on this line:

```
JSON marshal marshaling error for {TypeUrl:/cosmos.evidence.Equivocation Value:[8 20 18 11 8 128 146 184 195 152 254 255 255 255 1 24 100 34 20 49 173 204 219 253 249 89 153 48 188 152 56 119 184 172 10 201 144 167 165] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0 cachedValue:0xc0000aa480 compat:<nil>}, this is likely because amino is being used directly (instead of codec.Codec which is preferred) or UnpackInterfacesMessage is not defined for some type which contains a protobuf Any either directly or via one of its members. To see a stacktrace of where the error is coming from, set the var Debug = true in codec/types/compat.go
```

Not sure what to do here, should I use `v040Codec := codec.NewProtoCodec()` and register the interface/impls for Evidence & Equivocation?

alexanderbez replied:
for 0.40, is genesis state expected to be amino or proto encoded?

aaronc replied:
My general thought is that it should follow the chain configuration either amino or proto. We could have a flag to specify which format, but the simplest would be to just use the JSONMarshaler and TxConfig on `client.Context`.

alexanderbez replied:
> but the simplest would be to just use the JSONMarshaler and TxConfig on client.Context.

I would opt for this for 0.40 migration logic. The fields can be set on the `migrate` command flags.

aaronc replied:
The challenge now is with interface type registration is that right @amaurymartiny ?

amaurymartiny replied:
Interface registration is done in https://github.com/cosmos/cosmos-sdk/pull/6839/commits/766bb7ea3a2cc4e1cefd6cb7c1cdbd229ef35f23.

If the flag on `migrate` is optional, we could maybe add it in a follow-up PR?

amaurymartiny commented:
@alexanderbez yeah, your comments make sense, addressed.

> Has this been tested on exported state?

Not successfully, I was blocked by the OriginalVesting > Coins issue, will try again now.

alexanderbez commented:
> Not successfully, I was blocked by the OriginalVesting > Coins issue, will try again now.

That validation check needs to be removed. I removed it from `backport/0.39.x`.

amaurymartiny commented:
> That validation check needs to be removed. I removed it from backport/0.39.x.

Yeah, saw that. I also removed them here, in my latest couple of commits.

aaronc commented:
@amaurymartiny we will need to update this for `BaseAccount` pub key one #6928 is done FYI.

amaurymartiny commented on x/evidence/legacy/v0_40/genesis.pb.go:
```diff
@@ -0,0 +1,332 @@
+// Code generated by protoc-gen-gogo. DO NOT EDIT.
+// source: cosmos/evidence/genesis.proto
+
+package v040
```
I was finding myself manually copying a bunch of proto impls into `/legacy/v0_40/types.go`, including stuff like:
```golang
func (*GenesisState) Descriptor() ([]byte, []int) {
	return fileDescriptor_948832e7dde7529c, []int{0}
}
```

which is ugly. So I decided to just copy this whole file here.

amaurymartiny replied:
Note: for other modules (bank,slashing), only plain structs are being migrated, so I didn't copy their .pb.go files.

amaurymartiny commented on x/bank/legacy/v0_40/migrate.go:
```diff
@@ -2,21 +2,36 @@ package v040
 	v039auth "github.com/cosmos/cosmos-sdk/x/auth/legacy/v0_39"
+	v036supply "github.com/cosmos/cosmos-sdk/x/bank/legacy/v0_36"
 	v038bank "github.com/cosmos/cosmos-sdk/x/bank/legacy/v0_38"
+	v040bank "github.com/cosmos/cosmos-sdk/x/bank/types"
```
note to @alexanderbez: I remember you wanting the migrate scripts to be independent of the actual module's `types/` folder.

I initially went down this road, but it created some problems:
1. If I copied the `bank.pb.go` and `genesis.pb.go` files into `x/banks/legacy/v0_40`, then protobuf would show error "duplicate proto registration" because it registers each type twice, which pollutes the migrate command stdout.
2. If I copied some parts of `bank.pb.go` and `genesis.pb.go` into `x/banks/legacy/v0_40` (e.g. not copying the duplicate proto registration, or only copying the structs and some methods), that would involve a lot of work, and tweaking generated proto files, which is very prone to human errors.

So at Regen yesterday, we decided to just use the `types` from the modules:
- with these proto files, we're aiming at having very few breaking changes, which means that future migrate scripts can use the same proto definitions.
- if there's really a breaking change, then we'll copy the `.pb.go` into `legacy/` folders, but there won't be "duplicate proto registration" as we're registering under a new proto package name.

amaurymartiny commented on x/auth/legacy/v0_40/migrate.go:
```diff
@@ -1,24 +1,133 @@
 
+// convertBaseAccount converts a 0.39 BaseAccount to a 0.40 BaseAccount.
+func convertBaseAccount(old *v039auth.BaseAccount) *v040auth.BaseAccount {
+	// In `x/auth/legacy/v0_38/migrate.go`, when creating a BaseAccount, we
+	// explicitly set the PublicKey field to nil. This propagates until 0.40,
+	// and we don't know the PubKey for those accounts, so we just put an empty
+	// string inside the Any.
+	var any *codectypes.Any
+	var err error
+	if old.PubKey != nil {
+		any, err = tx.PubKeyToAny(old.PubKey)
+	} else {
+		s := pt.StringValue{Value: ""}
+		any, err = codectypes.NewAnyWithValue(&s)
+	}
```
Here's the output for BaseAccounts with nil pubkeys:

```js
     "auth": {
       "accounts": [
         {
           "@type": "/cosmos.auth.v1beta1.BaseAccount",
           "account_number": "0",
           "address": "cosmos1qqq9nqqllyvqfzen9w2sr0zp8jsfnycsjs8qwv",
           "pub_key": {
             "@type": "/google.protobuf.StringValue",
             "value": ""
           },
           "sequence": "1"
         },
// --snip--
```

is this okay? should I put an error message in the string value?

amaurymartiny replied:
Some context on the above comment.

In the current gaia hub, the genaccounts [don't have a `PubKey` field](https://github.com/cosmos/cosmos-sdk/blob/06b84d902a2038beef9927cc4eb1d72d5b1b35a9/x/genaccounts/legacy/v0_36/types.go#L14). When we convert that genesis to a more recent version (e.g. 0.37->0.38), the 0.38 migrate script can only create `BaseAccount` [with a nil `PubKey`](https://github.com/cosmos/cosmos-sdk/blob/6890feb3d28a1d98a8c8cd52669aca78361f5250/x/auth/legacy/v0_38/migrate.go#L16).

Continuing our genesis migration, we arrive at 0.40, where I show this protobuf String value in case there's no pubkey in the BaseAccount.

blushi replied:
what about using `google.protobuf.Empty`?

amaurymartiny replied:
Ah nice! I was looking around `Null` and `NullValue`, wasn't aware of empty!

amaurymartiny commented:
I tried updating a `genesis.json` from the hub3 (from ~2 months ago), and here are its successive migrated genesises: https://drive.google.com/drive/folders/1OQPJnx7PgyDtVi8A3z1ED4QThTnoPr2M

robert-zaremba reviewed (commented):
LGTM, but I don't know if we cover everything.

anilCSE commented on x/auth/legacy/v0_40/migrate.go:
```diff
@@ -1,24 +1,133 @@
+			{
+				v040Accounts[i] = &v040vesting.ContinuousVestingAccount{
+					BaseVestingAccount: convertBaseVestingAccount(v039Account.BaseVestingAccount),
+					StartTime:          v039Account.EndTime,
```
Should it be `v039Account.StartTime`?

anilCSE commented on x/auth/legacy/v0_40/migrate.go:
```diff
@@ -1,24 +1,133 @@
+			{
+				v040Accounts[i] = &v040vesting.ContinuousVestingAccount{
+					BaseVestingAccount: convertBaseVestingAccount(v039Account.BaseVestingAccount),
+					StartTime:          v039Account.EndTime,
```
Should it be v039Account.StartTime?

amaurymartiny replied:
Good catch, thanks a lot!

anilCSE commented on x/auth/legacy/v0_40/migrate.go:
```diff
@@ -1,24 +1,133 @@
+		case *v039auth.PeriodicVestingAccount:
+			{
+				vestingPeriods := make([]v040vesting.Period, len(v039Account.VestingPeriods))
+				for i, period := range v039Account.VestingPeriods {
```
`i` is ~~conflicting~~ confusing, use another variable

alessio commented on x/staking/legacy/v0_38/types.go:
```diff
@@ -1,6 +1,6 @@
 // DONTCOVER
 // nolint
-package v0_38
+package v038
```
Why not renaming the directory accordingly too?

alessio commented on x/genutil/types/types.go:
```diff
@@ -3,6 +3,7 @@ package types
 import (
 	"encoding/json"
 
+	"github.com/cosmos/cosmos-sdk/client"
```
This should show up after `"github.com/tendermint/tendermint/crypto"` (as per `goimports`)

alessio reviewed (approved):
utACK
